### PR TITLE
tools: api bench

### LINKED ISSUE #6922: Need API bench tools to test big cluster

### PR DESCRIPTION
**Pull request by CabinfeverB**
### What problem does this PR solve?

Issue Number: Close #6922

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
None.
```

### CODE REVIEW

File: tools/pd-api-bench/README.md [hunk 1]
```diff
@@ -0,0 +1,13 @@
+When test ScanRegions and GetRegion, it's better to set region-num.
+When test ScanRegions, it's better to set regions-sample, and not greater than 10000.
+``` bash
+go run ./main.go --region 10 --region-num 200000 --concurrency 10
+go run ./main.go --regions 10 --region-num 200000 --regions-sample 1000 --concurrency 10
```
Comment (rleungx) [inline tools/pd-api-bench/README.md:5]: What is the difference among `--regions`, `--region-num`, and `--regions-sample`?

Comment (CabinfeverB) [reply tools/pd-api-bench/README.md:5]: `ScanRegions` QPS, the number for total regions, the number of the result regions for a `ScanRegions`
I will modidy param name

File: tools/pd-api-bench/main.go [hunk 2]
```diff
@@ -0,0 +1,342 @@
+// Copyright 2022 TiKV Project Authors.
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:1]:
```suggestion
// Copyright 2023 TiKV Project Authors.
```

File: tools/pd-api-bench/main.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+	// GetStore qps
+	storeQps = flag.Int("qps-store", 0, "GetStore qps")
+	// GetStores qps
+	storesQps = flag.Int("qps-stores", 0, "GetStores qps")
+	// store max id
+	maxStoreID = flag.Int("max-store", 100, "store max id")
+	// concurrency
+	concurrency = flag.Int("concurrency", 1, "client number")
+	// brust
+	brust = flag.Int("brust", 1, "brust request")
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:58]:
```suggestion
	// burst
	burst = flag.Int("burst", 1, "burst request")
```

File: tools/pd-api-bench/main.go [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+	"syscall"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+)
+
+var (
+	pdAddr = flag.String("pd", "127.0.0.1:2379", "pd address")
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:37]: Do we need to support tls?

**Review by HuSharp — COMMENTED** [review]
btw, maybe need to add it in `Makefile`

File: tools/pd-api-bench/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+When test ScanRegions and GetRegion, it's better to set region-num.
+When test ScanRegions, it's better to set regions-sample, and not greater than 10000.
+``` bash
+go run ./main.go --qps-get-region 10 --region-num 200000 --concurrency 10
```
Comment (rleungx) [inline tools/pd-api-bench/README.md:4]: How about:
```suggestion
go run ./main.go --type get-region --qps 10 --region-num 200000 --concurrency 10
```

**Review by rleungx — APPROVED** [review]
Ci failed, the rest LGTM.

File: tools/pd-api-bench/README.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
+pd-api-bench
+========
+
+pd-api-bench is a tool to test PD API.
+
+## Build
+1. [Go](https://golang.org/) Version 1.20 or later
```
Comment (HuSharp) [inline tools/pd-api-bench/README.md:7]: please sync with the repo
```suggestion
1. [Go](https://golang.org/) Version 1.21 or later
```

File: tools/pd-api-bench/README.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+-pd string
+>  pd address (default "127.0.0.1:2379")
+
+-http-cases
+>  HTTP api bench cases list. Multiple cases are separated by commas. Such as `-http-cases GetRegionStatus,GetMinResolvedTS`
+>  You can set qps with '-' and set burst with '+' such as `-http-cases GetRegionStatus-100+1`
+
+-grpc-cases
+>  gRPC api bench cases list. Multiple cases are separated by commas.
```
Comment (HuSharp) [inline tools/pd-api-bench/README.md:50]: please add grpc usage as well

File: tools/pd-api-bench/README.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+>  print the output of api response for debug
+
+### TLS
+
+You can use the following command to generate a certificate for testing TLS:
+
+```shell
+mkdir cert
+./cert_opt.sh generate cert
+go run main.go --min-resolved-ts-http -cacert ./cert/ca.pem -cert ./cert/pd-server.pem  -key ./cert/pd-server-key.pem
```
Comment (HuSharp) [inline tools/pd-api-bench/README.md:68]: need to update this line

File: tools/pd-api-bench/go.mod [hunk 9]
```diff
@@ -0,0 +1,124 @@
+module github.com/tools/pd-api-bench
+
+go 1.20
```
Comment (HuSharp) [inline tools/pd-api-bench/go.mod:3]: ditto

File: tools/pd-api-bench/main.go [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+)
+
+var (
+	pdAddr    = flag.String("pd", "127.0.0.1:2379", "pd address")
+	debugFlag = flag.Bool("debug", false, "print the output of api response for debug")
+
+	httpCases = flag.String("http-cases", "", "http api cases")
+	gRPCCases = flag.String("grpc-cases", "", "grpc cases")
+
+	client = flag.Int("client", 1, "client number")
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:46]: Is it necessary to use a waitgroup to wait for all client calls to exit?

File: tools/pd-api-bench/main.go [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+	cases.PDAddress = fmt.Sprintf("%s://%s", protocol, addr)
+	cases.Debug = *debugFlag
+
+	hcases := make([]cases.HTTPCase, 0)
+	gcases := make([]cases.GRPCCase, 0)
+
+	var err error
+	hcaseStr := strings.Split(*httpCases, ",")
+	for _, str := range hcaseStr {
+		// fmt.Println(str)
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:90]: redundant line?

File: tools/pd-api-bench/main.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+	default:
+		exit(1)
+	}
+}
+
+func handleGRPCCase(ctx context.Context, gcase cases.GRPCCase, clients []pd.Client) {
+	log.Printf("begin to run gRPC case %s, with qps = %d and burst = %d", gcase.Name(), gcase.GetQPS(), gcase.GetBurst())
+	qps := gcase.GetQPS()
+	burst := gcase.GetBurst()
+	tt := base / qps * burst * int64(*client)
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:189]:
```suggestion
	qps := gcase.GetQPS()
	burst := gcase.GetBurst()
	tt := base / qps * burst * int64(*client)
	log.Printf("begin to run gRPC case %s, with qps = %d and burst = %d, rate is %d", gcase.Name(), qps, burst, tt)

```

File: tools/pd-api-bench/main.go [hunk 13]
```diff
@@ -0,0 +1,315 @@
...
+	qps   = flag.Int64("qps", 1000, "qps")
+	burst = flag.Int64("burst", 1, "burst")
+
+	// tls
+	caPath   = flag.String("cacert", "", "path of file that contains list of trusted SSL CAs")
+	certPath = flag.String("cert", "", "path of file that contains X509 certificate in PEM format")
+	keyPath  = flag.String("key", "", "path of file that contains X509 key in PEM format")
+)
+
+var base int64 = int64(time.Second) / int64(time.Microsecond)
```
Comment (HuSharp) [inline tools/pd-api-bench/main.go:57]: do you mean `Millisecond`?

Comment (CabinfeverB) [reply tools/pd-api-bench/main.go:57]: It is `Microsecond`, times  `Microsecond`  when calculating `tt`

**Review by HuSharp — APPROVED** [review]
LGTM!

**Pull request merged**